### ISSUE TITLE
Watch of *unstructured.Unstructured ended with: an error on the server

### ISSUE THREAD

**Issue by Gerrit-K** [pulumi/pulumi-kubernetes#1985]
### What happened?

***tl;dr***: my pulumi kubernetes deployment in combination with [loft-sh/kiosk](https://github.com/loft-sh/kiosk) or [HNS](https://github.com/kubernetes-sigs/hierarchical-namespaces) from a GitLab Agent prints several errors about failing watches, but without affecting the success of the overall deployment.

--

I'm currently participating in a poc testing phase for an (internal) platform provider, which aims to implement multi-tenancy on a managed kubernetes cluster via GitLab pipelines. Their current approach is based on restricting access to `Namespace`s and providing access to "lookalikes" via [loft-sh/kiosk](https://github.com/loft-sh/kiosk) or [HNS](https://github.com/kubernetes-sigs/hierarchical-namespaces) instead. The customer (me) then receives a read-only KUBECONFIG (to be used, for instance, on developer machines) and can deploy to the cluster via the GitLab Kubernetes Agent from within the CI/CD pipelines.

My first attempts to port my current product application to the new multi-tenancy environment failed and after a couple of days we tracked it down to a [rate-limiting issue with the GitLab Agent](https://gitlab.com/gitlab-org/gitlab/-/issues/354658), which eventually got resolved. One thing, however, we regularly see in our pipeline logs is this error:

```
Diagnostics:
  pulumi:pulumi:Stack ([REDACTED]):
    I0513 11:37:47.090687     155 request.go:665] Waited for 1.198921501s due to client-side throttling, not priority and fairness, request: POST:https://kas.gitlab.com/k8s-proxy/api/v1/namespaces/[REDACTED]/secrets
    W0513 11:38:40.543387     155 reflector.go:441] pkg/mod/k8s.io/client-go@v0.22.1/tools/cache/reflector.go:167: watch of *unstructured.Unstructured ended with: an error on the server ("unable to decode an event from the watch stream: stream error: stream ID 455; INTERNAL_ERROR") has prevented the request from succeeding
    W0513 11:38:40.591518     155 reflector.go:441] pkg/mod/k8s.io/client-go@v0.22.1/tools/cache/reflector.go:167: watch of *unstructured.Unstructured ended with: an error on the server ("unable to decode an event from the watch stream: stream error: stream ID 447; INTERNAL_ERROR") has prevented the request from succeeding
    W0513 11:38:40.630472     155 reflector.go:441] pkg/mod/k8s.io/client-go@v0.22.1/tools/cache/reflector.go:167: watch of *unstructured.Unstructured ended with: an error on the server ("unable to decode an event from the watch stream: stream error: stream ID 451; INTERNAL_ERROR") has prevented the request from succeeding
    W0513 11:39:41.569059     155 reflector.go:441] pkg/mod/k8s.io/client-go@v0.22.1/tools/cache/reflector.go:167: watch of *unstructured.Unstructured ended with: an error on the server ("unable to decode an event from the watch stream: stream error: stream ID 533; INTERNAL_ERROR") has prevented the request from succeeding
    W0513 11:39:41.966122     155 reflector.go:441] pkg/mod/k8s.io/client-go@v0.22.1/tools/cache/reflector.go:167: watch of *unstructured.Unstructured ended with: an error on the server ("unable to decode an event from the watch stream: stream error: stream ID 537; INTERNAL_ERROR") has prevented the request from succeeding
    W0513 11:39:42.743546     155 reflector.go:441] pkg/mod/k8s.io/client-go@v0.22.1/tools/cache/reflector.go:167: watch of *unstructured.Unstructured ended with: an error on the server ("unable to decode an event from the watch stream: stream error: stream ID 541; INTERNAL_ERROR") has prevented the request from succeeding
    W0513 11:40:47.917495     155 reflector.go:441] pkg/mod/k8s.io/client-go@v0.22.1/tools/cache/reflector.go:167: watch of *unstructured.Unstructured ended with: an error on the server ("unable to decode an event from the watch stream: stream error: stream ID 553; INTERNAL_ERROR") has prevented the request from succeeding
    W0513 11:41:48.817842     155 reflector.go:441] pkg/mod/k8s.io/client-go@v0.22.1/tools/cache/reflector.go:167: watch of *unstructured.Unstructured ended with: an error on the server ("unable to decode an event from the watch stream: stream error: stream ID 561; INTERNAL_ERROR") has prevented the request from succeeding
    W0513 11:41:48.856890     155 reflector.go:441] pkg/mod/k8s.io/client-go@v0.22.1/tools/cache/reflector.go:167: watch of *unstructured.Unstructured ended with: an error on the server ("unable to decode an event from the watch stream: stream error: stream ID 557; INTERNAL_ERROR") has prevented the request from succeeding
    W0513 11:41:48.873813     155 reflector.go:441] pkg/mod/k8s.io/client-go@v0.22.1/tools/cache/reflector.go:167: watch of *unstructured.Unstructured ended with: an error on the server ("unable to decode an event from the watch stream: stream error: stream ID 549; INTERNAL_ERROR") has prevented the request from succeeding
```

While I've never seen this error anywhere else (neither on my own testing AKS, nor on the currently-in-use production environment of said platform provider), it doesn't seem to cause any harm to the pulumi deployment. All resources are created and tracked correctly. It's just a bit worrying, so I was wondering if someone could explain where these errors come from (and maybe whom to turn to, if it's not an issue with the pulumi kubernetes provider :wink:). Thanks!

### Steps to reproduce

Due to the convoluted circumstances under which the error above occurs, I can't really give any precise reproduction steps 😕 

### Expected Behavior

No errors about failing watches

### Actual Behavior

Errors about failing watches (see description)

### Versions used

Output from `pulumi about`:

```
CLI          
Version      3.25.0
Go Version   go1.17.7
Go Compiler  gc
Plugins
NAME        VERSION
kubernetes  3.8.1
nodejs      unknown
random      4.3.1
terraform   5.4.1
Host     
OS       debian
Version  9.13
Arch     x86_64
This project is written in nodejs (/usr/local/bin/node v14.19.1)
Current Stack: [REDACTED]
Found no resources associated with [REDACTED]
Found no pending operations associated with [REDACTED]
Backend  
Name     runner-yabzwcfs-project-34025485-concurrent-09rzsr
URL      azblob://pulumi
User     root
NAME                VERSION
@pulumi/kubernetes  3.8.1
@pulumi/pulumi      3.15.0
@pulumi/random      4.3.1
@pulumi/terraform   5.4.1
pulumi-htpasswd     1.2.0
@types/node         14.17.27
prettier            2.4.1
Pulumi locates its logs in /tmp by default
warning: A new version of Pulumi is available. To upgrade from version '3.25.0' to '3.32.1', visit https://pulumi.com/docs/reference/install/ for manual instructions and release notes.
```

Output from `kubectl version`:
```
Client Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.7", GitCommit:"b56e432f2191419647a6a13b9f5867801850f969", GitTreeState:"clean", BuildDate:"2022-02-16T11:50:27Z", GoVersion:"go1.16.14", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.6", GitCommit:"07959215dd83b4ae6317b33c824f845abd578642", GitTreeState:"clean", BuildDate:"2022-03-30T18:28:25Z", GoVersion:"go1.16.12", Compiler:"gc", Platform:"linux/amd64"}
```

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by Gerrit-K** [comment]
Might be related to the gitlab agent after all, see [gitlab-agent/issues/255](https://gitlab.com/gitlab-org/cluster-integration/gitlab-agent/-/issues/255). I'll keep an eye on that and close this issue here if they are able to fix it there.

**Comment by Gerrit-K** [comment]
Sorry, @lblackstone , I honestly forgot about this ticket here while the other one was taking time to get resolved. I can confirm that it seems to be fixed by now and was most likely not related to pulumi-kubernetes. Thanks for closing it 👍
